vortex-serde/src/layouts/reader/builder.rs

```diff
@@ -0,0 +1,147 @@
+use crate::layouts::footer::Footer;
+use crate::layouts::reader::{DEFAULT_BATCH_SIZE, DEFAULT_PROJECTION, VortexLayoutBatchStream};
+use crate::layouts::reader::filtering::RowFilter;
+use crate::layouts::reader::projections::Projection;
+use crate::layouts::writer::layout_writer::MAGIC_BYTES;
+
+pub struct VortexLayoutReaderBuilder<R> {
+    reader: R,
+    layout_serde: LayoutReader,
+    projection: Option<Projection>,
+    len: Option<u64>,
+    indices: Option<Array>,
+    row_filter: Option<RowFilter>,
+    batch_size: Option<usize>,
+}
+
+impl<R: VortexReadAt> VortexLayoutReaderBuilder<R> {
+    // Recommended read-size according to the AWS performance guide
+    const FOOTER_READ_SIZE: usize = 8 * 1024 * 1024;
+    const FOOTER_TRAILER_SIZE: usize = 20;
+
+    pub fn new(reader: R, layout_serde: LayoutReader) -> Self {
+        Self {
+            reader,
+            layout_serde,
+            projection: None,
+            row_filter: None,
+            len: None,
+            indices: None,
+            batch_size: None,
+        }
+    }
+
+    pub fn with_length(mut self, len: u64) -> Self {
+        self.len = Some(len);
+        self
+    }
+
+    pub fn with_projection(mut self, projection: Projection) -> Self {
+        self.projection = Some(projection);
+        self
+    }
+
+    pub fn with_indices(mut self, array: Array) -> Self {
+        // TODO(#441): Allow providing boolean masks
+        assert!(
+            array.dtype().is_int(),
+            "Mask arrays have to be integer arrays"
+        );
+        self.indices = Some(array);
+        self
+    }
+
+    pub fn with_row_filter(mut self, row_filter: RowFilter) -> Self {
+        self.row_filter = Some(row_filter);
+        self
+    }
+
+    pub fn with_batch_size(mut self, batch_size: usize) -> Self {
+        self.batch_size = Some(batch_size);
+        self
+    }
+
+    pub async fn build(mut self) -> VortexResult<VortexLayoutBatchStream<R>> {
+        let footer = self.read_footer().await?;
+        let projection = self.projection.unwrap_or(DEFAULT_PROJECTION);
+        let batch_size = self.batch_size.unwrap_or(DEFAULT_BATCH_SIZE);
+
+        let scan = Scan {
+            projection,
+            indices: self.indices,
+            filter: self.row_filter,
+            batch_size,
+        };
+
+        let message_cache = Arc::new(RwLock::new(MessagesCache::default()));
+        let layouts_cache =
+            RelativeMessageCache::new(footer.dtype()?, message_cache.clone(), Vec::new());
+
+        let layout = footer.layout(scan.clone(), layouts_cache)?;
+
+        VortexLayoutBatchStream::try_new(self.reader, layout, message_cache, footer.dtype()?, scan)
+    }
+
+    async fn len(&self) -> usize {
+        let len = match self.len {
+            Some(l) => l,
+            None => self.reader.size().await,
+        };
+
+        len as usize
+    }
+
+    async fn read_footer(&mut self) -> VortexResult<Footer> {
```

Comment:
nit - we should probably split the IO and decoding logic into two separate functions, but that might be out of scope here

Reply:
I think we can change this later